src/libbson/src/bson/bson-dsl.h

```diff
@@ -0,0 +1,1320 @@
+#ifndef BSON_BSON_DSL_H_INCLUDED
+#define BSON_BSON_DSL_H_INCLUDED
+
+/**
+ * @file bson-dsl.h
+ * @brief Define a C-preprocessor DSL for working with BSON objects
+ *
+ * This file defines an embedded DSL for working with BSON objects consisely and
+ * correctly.
+ *
+ * For more information about using this DSL, refer to `bson-dsl.md`.
+ */
+
+#include "bson.h"
+
+enum {
+   /// Toggle this value to enable/disable debug output for all bsonDSL
+   /// operations (printed to stderr). You can also set a constant
+   /// BSON_DSL_DEBUG within the scope of a DSL command to selectively debug
+   /// only the commands within that scope.
+   BSON_DSL_DEBUG = 0
+};
+
+#ifdef _WIN32
+#define _bson_thread_local_comdat __declspec(thread) __declspec(selectany)
+#else
+#define _bson_thread_local_comdat __attribute__ ((weak)) __thread
+#endif
+
+#ifdef __GNUC__
+// GCC has a bug handling pragma statements that disable warnings within complex
+// nested macro expansions. If we're GCC, just disable -Wshadow outright:
+BSON_IF_GNU_LIKE (_Pragma ("GCC diagnostic ignored \"-Wshadow\""))
+#endif
+
+#define _bsonDSL_disableWarnings()                     \
+   BSON_IF_GNU_LIKE (_Pragma ("GCC diagnostic push");) \
+   BSON_IF_GNU_LIKE (_Pragma ("GCC diagnostic ignored \"-Wshadow\"");)
+
+#define _bsonDSL_restoreWarnings() \
+   BSON_IF_GNU_LIKE (_Pragma ("GCC diagnostic pop");)
+
+/**
+ * @brief Parse the given BSON document.
+ *
+ * @param doc A bson_t object to walk. (Not a pointer)
+ */
+#define bsonParse(Document, ...)                              \
+   _bsonDSL_begin ("bsonParse(%s)", _bsonDSL_str (Document)); \
+   _bsonDSL_disableWarnings ();                               \
+   bool _bvHalt = false;                                      \
+   bsonParseError = NULL;                                     \
+   const bool _bvContinue = false;                            \
+   const bool _bvBreak = false;                               \
+   (void) _bvContinue;                                        \
+   (void) _bvBreak;                                           \
+   (void) _bvHalt;                                            \
```

Comment:
```suggestion
   BSON_MAYBE_UNUSED const bool _bvContinue = false;          \
   BSON_MAYBE_UNUSED const bool _bvBreak = false;             \
   BSON_MAYBE_UNUSED bool _bvHalt = false;                    \
   bsonParseError = NULL;                                     \
```
Suggest use of `BSON_MAYBE_UNUSED` or `BSON_UNUSED` instead of explicit cast-to-void. Suggestion applies to other instances of cast-to-void in this file.

Reply:
I was unaware we had that macro.